### PR TITLE
[ROS-426] Port existing Fixnum primitives to current code base

### PR DESCRIPTION
**Pull request by AbnerZheng**
## Overview
Port existing Fixnum primitives to current code base

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/ROS-426

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

**Comment by AbnerZheng** [comment]
@tymm PTAL

File: roscala/macros/src/main/scala/coop/rchain/roscala/macros/Prim.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import scala.language.experimental.macros
+import scala.annotation.{compileTimeOnly, StaticAnnotation}
+import scala.reflect.macros.whitebox
+
+@compileTimeOnly("enable macro paradise to expand macro annotations")
+class checkTypeMismatch[A] extends StaticAnnotation {
+  def macroTransform(annottees: Any*): Any = macro TypeMismatchMacro.impl[A]
+}
+
+object TypeMismatchMacro {
```
Comment (AbnerZheng) [inline roscala/macros/src/main/scala/coop/rchain/roscala/macros/Prim.scala]: @tymm Should we use TypeMismatchMacro to check the type of arguments?

File: roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala [hunk 2]
```diff
@@ -5,13 +5,20 @@ import coop.rchain.roscala.Vm.State
...
 
 abstract class Prim extends Ob {
   def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob
 
-  // TODO: Add error case
-  def dispatchHelper(state: State, globalEnv: GlobalEnv): Ob =
-    fn(state.ctxt, globalEnv)
+  def dispatchHelper(state: State, globalEnv: GlobalEnv): Ob = {
+    val n = state.ctxt.nargs
+    if (minArgs <= n && n <= maxArgs)
```
Comment (AbnerZheng) [inline roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala]: I'm doing the checking of number of argument here like rosette do. Should I use macro as before?

File: roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala [hunk 3]
```diff
@@ -363,5 +371,45 @@ object Prim {
...
+    } else if (minArgs == maxArgs) {
+      if (minArgs == 1) {
+        "expected 1 argument"
+      } else {
+        s"expected $minArgs arguments"
+      }
+    } else {
+      s"expected between $minArgs and $maxArgs arguments"
+    }
+    return runtimeError(state.ctxt, msg)
```
Comment (AbnerZheng) [inline roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala]: We defined  some case class representing error like below:

```scala
sealed trait PrimError
case class ArgumentMismatch(msg: String)               extends PrimError
case class TypeMismatch(argNum: Int, typeName: String) extends PrimError
case class IndexOutOfBounds(argNum: Int, size: Int)    extends PrimError
case object ArithmeticError                            extends PrimError
```
I choose to do the way as rosette did. I'm not clearly here, can you tell me which way is preferred.

File: roscala/src/main/scala/coop/rchain/roscala/prim/fixnum.scala [hunk 4]
```diff
@@ -13,4 +21,369 @@ object fixnum {
...
+    @checkTypeMismatch[Fixnum]
+    override def fn(ctxt: Ctxt, globalEnv: GlobalEnv): Ob = {
+      val m = ctxt.argvec.value.head.asInstanceOf[Fixnum]
+      val n = ctxt.argvec.value(1).asInstanceOf[Fixnum]
+
+      m >>> n
+    }
+  }
+
+  private def checkFixnum(ctxt: Ctxt, n: Int, elem: Array[Ob]): Ob =
```
Comment (AbnerZheng) [inline roscala/src/main/scala/coop/rchain/roscala/prim/fixnum.scala]: Can you give me some advise of how to implement the checkFixnum?

File: roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala (hunk 2 above)
Comment (tymm) [reply roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala]: Seems like the macro is not really required.

File: roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala (hunk 3 above)
Comment (tymm) [reply roscala/src/main/scala/coop/rchain/roscala/prim/Prim.scala]: Yes, let's stay close to Rosette. If you discover already translated primitives that use `PrimError` let me know and we will decide what to do.

File: roscala/src/main/scala/coop/rchain/roscala/prim/fixnum.scala (hunk 4 above)
Comment (tymm) [reply roscala/src/main/scala/coop/rchain/roscala/prim/fixnum.scala]: It looks correct to me.
If `elem(n)` is no `Fixnum` it does some error handling stuff and then returns `DEADTHREAD`.
If not, it returns `elem(n)` as a `Fixnum`.

**Pull request merged**